File: Wff.Desktop/ViewModels/MainWindowViewModel.fs
namespace Wff.ViewModels
open Utils

open CommunityToolkit.Mvvm
open Avalonia
open Avalonia.Platform.Storage
open System.Threading.Tasks
open Utils.Recorder
open System.Collections.Generic
open System

type MainWindowViewModel() as self =
    inherit ViewModelBase()
    //Backing fields
    let framerates = [ "Default"; "30"; "60"; "120" ]
    let audioBackends = [ "Default"; "Pipewire"; "PulseAudio"; "ALSA" ]
    let videoCodecs = Utils.ffmpeg.codecs Utils.ffmpeg.FfmpegCodecTarget.Video
    let audioCodecs = Utils.ffmpeg.codecs Utils.ffmpeg.FfmpegCodecTarget.Audio
    let audioDevices = Utils.ffmpeg.audioDevices
    let outputs = Utils.ffmpeg.outputs
    let mutable output = outputs.[0];
    let mutable WAYLAND_DISPLAY = None
    let mutable audioCodec = audioCodecs.[0]
    let mutable videoCodec = videoCodecs.[0]
    let mutable framerate = framerates.[0]
    let mutable cancelledDuringCountdown = false
    let mutable region = "Screen"
    let mutable delay = 0
    let mutable dmabuf = true
    let mutable recordingDuration = ""
    let mutable damage = true
    let mutable audioBackend = audioBackends.[0]
    let mutable audioDevice = audioDevices.[0]
    let mutable RecorderProcess: Utils.Recorder.RecorderProcess option = None
    let mutable CountdownProcess = None
    let mutable recording = false

    let mutable filename :string  =
        let wffFolder =
            [ Environment.GetFolderPath Environment.SpecialFolder.MyVideos
              "wff-recordings" ]
            |> List.toArray
            |> System.IO.Path.Join

        if System.IO.Directory.Exists wffFolder <> true then
            System.IO.Directory.CreateDirectory wffFolder |> ignore
        let ts = DateTime.Now.ToString("yyyyMMdd_HHmmss")
        [ wffFolder; $"{ts}output.mkv" ] |> List.toArray |> System.IO.Path.Join

    let timerElapsed (e: Timers.ElapsedEventArgs) =
        if RecorderProcess.IsSome then
            self.RecordingDuration <- RecorderProcess.Value.Stopwatch.Elapsed.ToString @"hh\:mm\:ss\:ff"

    let StartCountdownAsync() = async {      
            let _process = new System.Diagnostics.Process()
            CountdownProcess <- Some _process;
            let countdownExe = System.IO.Path.Combine(AppDomain.CurrentDomain.BaseDirectory,"countdown")
            _process.StartInfo <-
                System.Diagnostics.ProcessStartInfo(
                    FileName = countdownExe, 
                    Arguments = $"{self.Output} {self.Delay}",
                    RedirectStandardOutput = true,
                    RedirectStandardError = true,
                    CreateNoWindow = true,
                    UseShellExecute = false
                )

            _process.Start() |> ignore
            _process.BeginOutputReadLine()
            do! _process.WaitForExitAsync() |> Async.AwaitTask
        }
    let StartRecording (model: Utils.Recorder.WfrecorderDataModel) = async {
        self.IsRecording <- true
        cancelledDuringCountdown <- false
        if self.Delay > 0 then
            do! StartCountdownAsync() //will block till countdown is complete
        if cancelledDuringCountdown <> true then
            let recorderProcess = Utils.Recorder.CreateRecorder model
            CountdownProcess <- None
            RecorderProcess <- Some recorderProcess
            match recorderProcess.Start with
            | Error e -> printfn "%s" e
            | Ok _ ->
                let timer = new System.Timers.Timer 100 //ms
                timer.Start()
                timer.Elapsed.Add timerElapsed
                printfn "Starting recording..."
        } 
    let SaveFileDialog() =
        //capture the wayland display before it is overwritten by the portal
        let wayland_display = Environment.GetEnvironmentVariable("WAYLAND_DISPLAY")
        match Application.Current with
        | null -> ()
        | app ->
            match app.ApplicationLifetime with
            | :? Controls.ApplicationLifetimes.IClassicDesktopStyleApplicationLifetime as desktop ->
                match desktop.MainWindow with
                | null -> ()
                | mainWindow ->
                    let saveFileAsync = async {
                        let options = FilePickerSaveOptions()
                        options.SuggestedFileName <- System.IO.Path.GetFileName(self.Filename:string)                  
                        options.ShowOverwritePrompt <- true
                        options.Title <- "Save Recording"
                        let! file = mainWindow.StorageProvider.SaveFilePickerAsync options |> Async.AwaitTask
                        match file with
                        | null -> ()
                        | f -> self.Filename <- f.Path.AbsolutePath
                        //restore Environment
                        maybe {
                            let! var = wayland_display
                            Environment.SetEnvironmentVariable("WAYLAND_DISPLAY", var)
                        } |> ignore
                    }
                    Async.StartImmediate saveFileAsync
            | _ -> ()        

    let StopRecording () =
        //stop a recorder process along with a countdown process if they exist
        if self.IsRecording then
            match CountdownProcess with
                | None -> ()
                | Some p ->
                    cancelledDuringCountdown <- true
                    p.Signal 2 |> ignore
                    CountdownProcess <- None
            match RecorderProcess with
                | None -> ()
                | Some rp ->                
                    match rp.Stop with
                    | Error e -> printfn "%s" e
                    | Ok _ ->
                        rp.Dispose()
                        RecorderProcess <- None
                        printfn "Recording saved ..."
            self.IsRecording <- false

    let TrySlurpSelectRegion () =
        //Use slurp to capture a region of the screen
        let _process = new System.Diagnostics.Process()
        _process.StartInfo <-
            System.Diagnostics.ProcessStartInfo(
                FileName = "slurp", //TODO, need to check if slurp is installed before this is run
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                CreateNoWindow = true,
                UseShellExecute = false
            )

        _process.Start() |> ignore
        _process.BeginOutputReadLine()

        _process.OutputDataReceived.Add(fun data ->
            printfn "trying to run command"
            data.Data
            |> Option.ofObj
            |> Option.iter (fun d ->
                printfn "%s" d
                if d.Length > 0 then
                    self.Region <- d))
        _process.ErrorDataReceived.Add(fun e ->
            printfn "recieved error %s" e.Data
        )

        _process.WaitForExit()

    member self.AudioDevices = Utils.ffmpeg.audioDevices

    member self.AudioCodecs = audioCodecs

    member self.VideoCodecs = videoCodecs

    member self.Framerates = framerates

    member self.AudioBackends = audioBackends
    member self.Outputs = outputs

    member self.Filename
        with get () = filename
        and set v = self.SetProperty(&filename, v) |> ignore

    member self.Dmabuf
        with get () = dmabuf
        and set v = self.SetProperty(&dmabuf, v) |> ignore

    member self.Damage
        with get () = damage
        and set v = self.SetProperty(&damage, v) |> ignore

    member self.AudioBackend
        with get () = audioBackend
        and set v = self.SetProperty(&audioBackend, v) |> ignore

    member self.AudioCodec
        with get () = audioCodec
        and set v = self.SetProperty(&audioCodec, v) |> ignore

    member self.VideoCodec
        with get () = videoCodec
        and set v = self.SetProperty(&videoCodec, v) |> ignore

    member self.AudioDevice
        with get () = audioDevice
        and set v = self.SetProperty(&audioDevice, v) |> ignore

    member self.Framerate
        with get () = framerate
        and set v = self.SetProperty(&framerate, v) |> ignore

    member self.IsRecording
        with get () = recording
        and set v = self.SetProperty(&recording, v) |> ignore

    member self.TryStartRecordingCommand =
        CommunityToolkit.Mvvm.Input.RelayCommand(fun _ -> StartRecording self.DataModel |> Async.Start)

    member self.TryStopRecordingCommand =
        CommunityToolkit.Mvvm.Input.RelayCommand(fun _ -> StopRecording())

    member self.TrySlurpSelectRegionCommand =
        CommunityToolkit.Mvvm.Input.RelayCommand(fun _ -> TrySlurpSelectRegion())

    member self.TrySelectScreenRegionCommand =
        CommunityToolkit.Mvvm.Input.RelayCommand(fun _ -> self.Region <- "Screen")

    member self.SelectFile =
        CommunityToolkit.Mvvm.Input.RelayCommand(fun _ -> SaveFileDialog())

    member self.RecordingDuration
        with get () = recordingDuration
        and set v = self.SetProperty(&recordingDuration, v) |> ignore

    member self.Region
        with get () = region
        and set v = self.SetProperty(&region, v) |> ignore

    member self.Delay
        with get() = delay
        and set v = self.SetProperty(&delay, v) |> ignore

    member self.Output
        with get() = output
        and set v = self.SetProperty(&output, v) |> ignore

    member private self.DataModel: Utils.Recorder.WfrecorderDataModel =
        { Framerate = self.Framerate
          Filename = self.Filename
          Output = self.Output  
          VideoCodec = self.VideoCodec
          AudioCodec = self.AudioCodec
          Dmabuf = self.Dmabuf
          Damage = self.Damage
          Region = self.Region
          AudioBackend = self.AudioBackend
          AudioDevice = self.AudioDevice }

File: Wff.Desktop/ViewModels/MainWindowViewModel.cs
namespace Wff.Desktop.ViewModels;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Threading.Tasks;
using Avalonia;
using Avalonia.Platform.Storage;
using CommunityToolkit.Mvvm.ComponentModel;
using Wff.Utils;

public partial class MainWindowViewModel : ViewModelBase
{
    public MainWindowViewModel()
    {
        _ = InitAsync();
    }
    List<string> Framerates => new() { "Default", "30", "60", "120" };
    List<string> audioBackends => new() { "Default", "Pipewire", "Pulseaudio", "ALSA" };
    RecorderProcess? recorderProcess = null;
    Process? countdownProcess = null;
    bool cancelledDuringCountdown = false;
    string initialFilename()
    {
        string wffFolder = Path.Join(
            Environment.GetFolderPath(Environment.SpecialFolder.MyVideos),
            "wff-recordings"
        );
        if (!Directory.Exists(wffFolder))
        {
            Directory.CreateDirectory(wffFolder);
        }
        string ts = DateTime.Now.ToString("yyyyMMdd_HHmmss");
        return Path.Join(wffFolder, $"{ts}output.mkv");
    }

    void TimerElapsed(object? s, System.Timers.ElapsedEventArgs e)
    {
        if (recorderProcess is not null)
        {
            RecordingDuration = recorderProcess.stopwatch.Elapsed.ToString(@"hh\:mm\:ss\:ff");
        }
    }

    async Task StartCountdownAsync()
    {
        var process = new Process();
        countdownProcess = process;
        var countdownExe = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "countdown");
        Console.WriteLine($"{countdownExe} {Output} {Delay}");
        process.StartInfo = new()
        {
            FileName = countdownExe,
            Arguments = $"{Output} {Delay}",
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            CreateNoWindow = true,
            UseShellExecute = false
        };
        if (process is null) return;
        process.Start();
        process.BeginOutputReadLine();
        await process.WaitForExitAsync();
    }

    public async void StartRecordingAsync()
    {
        IsRecording = true;
        cancelledDuringCountdown = false;
        if (Delay > 0)
        {
            await StartCountdownAsync();
        } //will block till countdown is complete

        if (cancelledDuringCountdown) return;
        recorderProcess = Wfrecorder.CreateRecorder(DataModel);
        recorderProcess.Start();
        var timer = new System.Timers.Timer(100);
        timer.Start();
        timer.Elapsed += TimerElapsed;
        Console.WriteLine("Starting recording...");
    }
    public async void SaveFileDialogAsync()
    {
        var wayland_display = Environment.GetEnvironmentVariable("WAYLAND_DISPLAY");
        var toplevel = (Application.Current?.ApplicationLifetime as Avalonia.Controls.ApplicationLifetimes.IClassicDesktopStyleApplicationLifetime)?.MainWindow;
        var options = new FilePickerSaveOptions();
        options.SuggestedFileName = Path.GetFileName(Filename);
        options.ShowOverwritePrompt = true;
        options.Title = "Save Recording";
        if (toplevel is null) return;
        var file = await toplevel.StorageProvider.SaveFilePickerAsync(options);
        if (file is not null) Filename = file.Path.AbsolutePath;
        //restore environment
        if (wayland_display is not null)
        {
            Environment.SetEnvironmentVariable("WAYLAND_DISPLAY", wayland_display);
        }
    }

    public void TrySelectScreenRegion() => Region = "Screen";

    public void StopRecording()
    {
        //stop a recorder process along with a countdown process if they exist
        if (!IsRecording) return;
        if (countdownProcess is not null)
        {
            cancelledDuringCountdown = true;
            countdownProcess.Signal(2);
            countdownProcess = null;
        }
        if (recorderProcess is not null)
        {
            recorderProcess.Stop();
            recorderProcess = null;
            Console.WriteLine("Recording saved");
        }
        IsRecording = false;
    }

    public void TrySlurpSelectRegion()
    {
        var process = new Process();
        process.StartInfo = new()
        {
            FileName = "slurp",
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            CreateNoWindow = true,
            UseShellExecute = false
        };
        if (process is null) return;
        process.Start();
        process.BeginOutputReadLine();
        process.OutputDataReceived += (s, e) =>
        {
            if (e.Data is null) return;
            var data = e.Data;
            if (data.Length > 0)
            {
                Region = data;
            }
        };
        process.WaitForExit();
    }
    [ObservableProperty]
    public string filename = "";

    [ObservableProperty]
    public bool dmabuf = true;

    [ObservableProperty]
    public bool damage = true;

    [ObservableProperty]
    public string audioBackend = "";

    [ObservableProperty]
    public string audioCodec = "";

    [ObservableProperty]
    public string videoCodec = "";

    [ObservableProperty]
    public string audioDevice = "";

    [ObservableProperty]
    public string framerate = "";

    [ObservableProperty]
    public bool isRecording = false;

    [ObservableProperty]
    public string recordingDuration = "";

    [ObservableProperty]
    public string region = "Screen";

    [ObservableProperty]
    public int delay = 0;

    [ObservableProperty]
    public string output = "";

    public async Task InitAsync()
    {
        var audioDevicesTask = Utils.Ffmpeg.AudioDevicesAsync();
        var audioCodecsTask = Utils.Ffmpeg.CodecsAsync(Utils.FfmpegCodecTarget.Audio);
        var videoCodecsTask = Utils.Ffmpeg.CodecsAsync(Utils.FfmpegCodecTarget.Video);
        var outputsTask = Utils.Ffmpeg.OutputsAsync();
        await Task.WhenAll(audioDevicesTask, audioCodecsTask, videoCodecsTask, outputsTask);

        AudioDevices = await audioDevicesTask;
        AudioCodecs = await audioCodecsTask;
        VideoCodecs = await videoCodecsTask;
        Outputs = await outputsTask;
        Filename = initialFilename();
        Output = Outputs[0];
        AudioDevice = AudioDevices[0];
        AudioCodec = AudioCodecs[0];
        VideoCodec = VideoCodecs[0];
        Framerate = Framerates[0];
        AudioBackend = AudioBackends[0];
        Console.WriteLine(Outputs.Count);
    }
    [ObservableProperty]
    public List<string> audioDevices = new();
    [ObservableProperty]
    public List<string> audioCodecs = new();
    [ObservableProperty]
    public List<string> videoCodecs = new();
    [ObservableProperty]
    public List<string> outputs = new();
    public List<string> AudioBackends { get; set; } = new() { "Default", "Pipewire", "Pulseaudio", "ALSA" };
    WfrecorderDataModel DataModel => new()
    {
        Framerate = Framerate,
        Filename = Filename,
        Output = Output,
        VideoCodec = VideoCodec,
        AudioCodec = AudioCodec,
        Dmabuf = Dmabuf,
        Damage = Damage,
        Region = Region,
        AudioBackend = AudioBackend,
        AudioDevice = AudioDevice,
    };
}
